Fix Bug: Regular-User validation webhook is treating SREs as regular users

jewzaam opened the pull request:
https://issues.redhat.com/browse/OSD-3242

SRE is now allowed to modify the resources the regular user validation webhook gates.
Added unit tests for regular_user_validation (yay!)
Added generation of requirements.txt (just in case you need it)

lisa commented on Makefile:
```diff
@@ -24,13 +23,25 @@ CONTAINER_ENGINE?=docker
 
+.PHONY: clean
+clean:
+	$(CONTAINER_ENGINE) rmi $(REPO_NAME):test || true
+	$(CONTAINER_ENGINE) rmi $(IMG):$(IMAGETAG) || true
```
```suggestion
	$(CONTAINER_ENGINE) rmi $(REPO_NAME):test $(IMG):$(IMAGETAG) || true
```

jewzaam replied:
Torn on this, I like using something different simply because it is going to be harder to accidentally push since it's not setup for that.  It isn't something we should be thinking we can push.. thoughts?

lisa replied:
> Torn on this, I like using something different simply because it is going to be harder to accidentally push since it's not setup for that. It isn't something we should be thinking we can push.. thoughts?

I'm not sure I follow. This is the `clean` target, where we want to remove the exiting images. The two `rmi` can be condensed into a single one. Am I missing something?

jewzaam replied:
ack, somehow I missed what the change was.. merging thanks

lisa commented on Makefile:
```diff
@@ -24,13 +23,25 @@ CONTAINER_ENGINE?=docker
 default: all
 all: build-base build-sss
```
Is it worth looping in the `test` target here?

jewzaam replied:
It's pulled in as a dependency by build-base.

jewzaam commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,33 @@ def handle_request():
+    return responses.response_invalid()
+
+
+def is_request_allowed(username, groupname, admin_groupnames=[]):
+  if groupname not in admin_groupnames and (username == "system:unauthenticated" or (not username.startswith("kube:") and not username.startswith("system:"))):
+    return False
+  else:
+    return True
```
Probably could.  My initial try didn't work so I just did this to get it done..

2uasimojo replied:
IMHO it would be more scrutable if you split it up _further_.
```
if groupname in admin_groupnames:
    return True
if username == "system:authenticated":
    return True
if username.split(":", 1)[0] in ("kube", "system"):
    return True
return False
```

jewzaam replied:
Pushed an update to `return not(<the if condition>)`.
Suggestions welcome, but I think it's reasonable to be able to read it as it is.

2uasimojo commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,33 @@ def handle_request():
+    else:
+      group_name = body_dict['object']['metadata']['name']
+
     # get the username and decide if it's a special user (SA, kube:admin, etc) or not.. 
     # deny any request if it's not a special user:  kube:*, system:* except system:unauthenticated
+    # ALLOW any request if the group_name is in admin_groups
     # reference: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request
```
Consider moving this comment into `is_request_allowed` itself. (Perhaps even as a docstring.)

jewzaam replied:
I'm weak on my python-foo, suggestions for the changes?

2uasimojo replied:
See below.

2uasimojo commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,30 @@ def handle_request():
-    body_dict = request.json['request']
-    username = body_dict['userInfo']['username']
-    if username == "system:unauthenticated" or (not username.startswith("kube:") and not username.startswith("system:")):
+    if is_request_allowed(username, group_name, admin_groups):
```
Wait wait, isn't this backwards now?
```
if is_request_allowed():
    return responses.response_deny()
else:
    return responses.response_allow()
```

jewzaam replied:
Yup, pushed a fix with unit tests.  Thanks!

2uasimojo commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,30 @@ def handle_request():
+    else:
+      group_name = body_dict['object']['metadata']['name']
+
     # get the username and decide if it's a special user (SA, kube:admin, etc) or not.. 
     # deny any request if it's not a special user:  kube:*, system:* except system:unauthenticated
+    # ALLOW any request if the group_name is in admin_groups
     # reference: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request
```
Suggestion part 1. The details of how you're deciding whether the request should be allowed are now within `is_request_allowed` and don't need to be here.
```suggestion
```

2uasimojo commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,30 @@ def handle_request():
+    return responses.response_invalid()
+
+
+def is_request_allowed(username, groupname, admin_groupnames=[]):
```
Suggestion part 2, use a docstring for the method and include the reference.
```suggestion
def is_request_allowed(username, groupname, admin_groupnames=()):
    """Decide if it's a special user (SA, kube:admin, etc) or not.

    reference: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request
    """
```

2uasimojo commented on src/webhook/regular_user_validation.py:
```diff
@@ -25,22 +31,30 @@ def handle_request():
+
+
+def is_request_allowed(username, groupname, admin_groupnames=[]):
+  return not(groupname not in admin_groupnames and (username == "system:unauthenticated" or (not username.startswith("kube:") and not username.startswith("system:"))))
```
Yeah, this is way hard to understand IMO, and very not-pythonic.
Suggestion part 3:
```suggestion
  # Explicitly unauthenticated
  if username == "system:unauthenticated":
      return False
  if groupname in admin_groupnames:
      return True
  if username.startswith("kube:") or username.startswith("system:"):
      return True
  # Deny anyone else
  return False
```
If you do this, you don't need the comments removed from above at all:
```
    # deny any request if it's not a special user:  kube:*, system:* except system:unauthenticated
    # ALLOW any request if the group_name is in admin_groups
```
...because that's obvious from the code.

jewzaam commented:
Think all things are resolved now.  Refactored request handling to a separate function and added unit tests.  It did indeed fail the tests until logic was fixed.  Thanks @2uasimojo !

2uasimojo commented on src/webhook/test_regular_user_validation.py:
```diff
@@ -1,23 +1,70 @@
 ]
 
+def create_request(uid, kind, username, groupname):
+    request = {
+        "json": {
+            "request": {
+                "uid": "testuser",
+                "operation": "UPDATE",
+                "userInfo": {
+                    "username": username
+                },
+                "oldObject": {
+                    "metadata": {
+                        "name": groupname
+                    }
+                },
+                "object": {
+                    "kind": kind,
+                    "metadata": {
+                        "name": groupname
+                    }
+                },
+            }
+        }
+    }
+    return request
```
Following on from above, I think you could avoid having to change the way you're accessing `request.json` by making this:

```suggestion
    class FakeRequest(object):
        json = {
            "request": {
                "uid": "testuser",
                "operation": "UPDATE",
                "userInfo": {
                    "username": username
                },
                "oldObject": {
                    "metadata": {
                        "name": groupname
                    }
                },
                "object": {
                    "kind": kind,
                    "metadata": {
                        "name": groupname
                    }
                },
            }
        }

    return FakeRequest()
```

2uasimojo commented on src/webhook/test_regular_user_validation.py:
```diff
@@ -1,5 +1,6 @@
 import unittest
 import json
+import enum
```
Confirmed...
```
[efried@efried managed-cluster-validating-webhooks]$ flake8 --select=F401 src/webhook/test_regular_user_validation.py 
src/webhook/test_regular_user_validation.py:3:1: F401 'enum' imported but unused
```
...but this isn't the only pep8 error in this file
```
[efried@efried managed-cluster-validating-webhooks]$ flake8 src/webhook/test_regular_user_validation.py  | wc -l
11
```
or in this repo
```
[efried@efried managed-cluster-validating-webhooks]$ find . -name '*.py' | xargs flake8 | wc -l
242
```
so probably not something to attack in this PR.

Could consider a future PR to add flake8 to the CI though...

2uasimojo commented:
I'd feel better if @lisa gave this a nod as well before it was merged, but
/lgtm
